### ISSUE TITLE
Mouse Selection Issue

### ISSUE THREAD

**Issue by RabidPistachio** [jgmenu/jgmenu#189]
**jgmenu version: 4.4.1**

Hello. I have been messing around with jgmenu and noticed that if you have widgets with actions on the menu it will do the action of the last hovered over widget when the mouse is clicked. Even if I have selected a menu item. Keyboard inputs work fine though. I did some digging in the code and found the issue in jgmenu.c, specifically this line:

`1877		if (ret && ret[0] != '\0') {`

This line is part of this code snippet on the mouse_function shown partially below for more context:

```
1870        /* left-click */
1871	if (ev->button == Button1) {
1872		char *ret;
1873
1874		/* widgets */
1875		widgets_set_pointer_position(mouse_coords.x, mouse_coords.y);
1876		ret = widgets_get_selection_action();
1877		if (ret && ret[0] != '\0') {
1878			action_cmd(ret, NULL);
1879			return;
1880		}
1881
1882		/* normal menu items */
```

What the code does is when the left mouse button is released, it first updates the selected widget where the mouse is pointing. Then it executes the action if the widget has a valid action and then return out of the function, ignoring the menu items. This behavior is fine when you are selecting a widget and the program works correctly when selecting between widgets. However the problem comes in when you click outside of a widget. The code above executes again with widgets_set_pointer_position() not updating the selection variable from widgets.c but mouseover is set to 0 due to not being over any widget. The next line (1876) also executes using the selection of the last moused over widget and should return an action. If you notice, line 1877 checks if there is a valid action to perform and does not check to see if the mouse is actually over the widget. So because of that, if there is an least one actionable widget in the menu, then selection should always be a valid widget and thus line 1877 should always be true and return out of the function doing the widget's action and not ever getting to the menu items. This was easily fixed by checking if the mouseover variable was set to 1 since that is reset when widgets_set_pointer_position() is called. There is probably a more elegant way to do this but this is what I did:

`1877		if (widgets_mouseover() == 1 && ret && ret[0] != '\0') {`

This should mean that now the if statement should be true if there is a valid action and the mouse is over a widget. In my very short testing I was able to select both menu items and actionable widgets with the mouse like I would expect. If you would like an example, I was testing with the files that I copied from the link below as a guide, it showcases the behavior:

https://forums.bunsenlabs.org/viewtopic.php?id=6540

Or  you can make a file with some basic menu items and a widget that opens a program when clicked on and see if you can get anything to open besides that program.

**tl;dr**

Mouse Release event doesn't check if it's actually over the last selected widget and executes actions anyway. Potential fix:

`1877		if (widgets_mouseover() == 1 && ret && ret[0] != '\0') {`

**Comment by johanmalm** [comment]
Thanks for identifying the bug and doing the analysis.
I've implemented.
